### ISSUE TITLE
Hint if one of the multiple types could not have a function

### ISSUE THREAD

**Issue by panvid** [kalessil/phpinspectionsea#813]
I like information, that there is a possible `null pointer` on calling a function on a possible `null`. The same I need in following case:

```
class Foo1{}

class Foo2
{
    public function foo(){}
}

/**
 * @return  Foo1|Foo2
 */
function bar(){}

$bar = bar();
$bar->foo();
```

Function `bar()` can return a `Foo1` or `Foo2`. The function `foo()` only exists in Foo2, so the call `$bar->foo()` should show a hint like:

> Method `foo` not found in every possible objects.

**Comment by rentalhost** [comment]
I think it valid. But what you are trying to do is not "impossible", but is not a good programming code. If `Foo1` is not like `Foo2`, so you should not mix types here.

But, as I said, I think it valid once that this inspection should not validate the code quality, but if it will give an error or not, and this case is probably that we found an error on runtime.

So I think that it should be implemented, but we should think too to how avoid false-positives. And I think that it should be valid:

```php
/** @var Foo1|Foo2 $bar */
if ($bar instanceof Foo1) {
    // Ok: as now $bar is strictly Foo1.
    $bar->foo();
}

// Warning: foo() doesn't exists on Foo2.
$bar->foo();
```

**Comment by kalessil** [comment]
My beloved PhpDoc limitations =)

Wouldn't it be better to check if combining classes in PhpDoc is making sense - I understand the case, but it's also absolutely expected outcome?

**Comment by dpauli** [comment]
Its totally correct, that this usecase is not a *good programming code* :) I work with a > 3 year old code, which is started in PHP 5.x and without totally OOP, so a refactoring would be better, correct.

I also can _reproduce_ this problem with cleaner real live code:
```php
class SomeAbstract
{
  proctected $id;

  abstract public function foo(): SomeAbstract;

  public function getId()
  {
    return $this->id;
  }
}

class Bar1 extends SomeAbstract
{
  protected $name;

  /**
   * @return Bar1
   */
  public function foo(): SomeAbstract {
    return $this;
  }

  public function getName()
  {
    return $this->name;
  }
}

class Bar2 extends SomeAbstract
{
  /**
   * @return Bar2
   */
  public function foo(): SomeAbstract
  {
    return $this;
  }
}

/**
  * @return Bar1|Bar2
  */
public function foo(): SomeAbstract
{
  ...
}

$foo = foo();
// works because getId() exists in abstract
$foo->getId();

// should show a hint because getName() only exist in Bar1
$foo->getName();
```

For me it sounds valid to use OOP like this: some function can return a instance from SomeAbstract, which can have their own functions.

**Comment by kalessil** [comment]
I'm not judging anything (IDE does xD). But speaking seriously, what will be more efficient for this specific context:
- checking that both `Bar1|Bar2` has needed method
- checking PhpDoc

To clarify: I can do only one of them and want to be super-practical here =)

**Comment by dpauli** [comment]
I think the first idea is the best: checking if both classes has the needed method.

So structs like
```
/** @var Foo1|Foo2 $bar */
if ($bar instanceof Foo1) {
    // Ok: as now $bar is strictly Foo1.
    $bar->foo();
}
```
help avoid this problem.

**Comment by kalessil** [comment]
@rentalhost : any other cases (except instanceof) we should be aware of?

**Comment by rentalhost** [comment]
We have a lot of possible cases like that, but I don't think feasible handle all for now. Eg. is_a(), get_class()...

**Comment by funivan** [comment]
Hello everyone. 
In this use case, it will be better to create `interface` and change return type. Or... create own wrapper to make code more clean (php 7, return types etc)

Case is common but maybe it will be more useful to introduce "Inconsistent type declaration"
We can have multiple return types declared in following cases:
- function return type
- function argument type
- property type
- variable var doc

`Inconsistent type declaration` can catch this cases. The inspection will check `PhpDoc` and warn if we have multiple types. 
```php
class UserList {
/**
 * @var SqlDb|JsonDb  <- we should place warning here
 */
private $storage;
}
```

```php
/**
 * @return  Foo1|Foo2  <-- warn here
 */
function bar(){}
```

We can skep types with null:
```
string|null    <-- ok
Foo1|null  <-- ok
Foo1|Foo2|null  <-- warning
```

I`m planning to implement this inspection but don't know if this will be useful for everyone. It can catch @davidmpaz your cases and help.

@kalessil if this is interesting I can implement it in Ultimate or Community edition.

**Comment by dpauli** [comment]
Its also okay for me to have the hint, that this function can return multiple (not depending) types. But mention following struct, which can work (and still work in our code):

```
class SomeAbstract
{
  /**
   * @return AbstractModel
   */
  public function foo()
  {
  }
}

class SomeImplementation extends SomeAbstract
{
  /**
   * @return ImplementationModel|AbstractModel
   */
  public function foo()
  {
    $model = self::foo(); // this returns an AbstractModel and do some parent stuff
    // do other, implementation, stuff
    return $model;
  }
}
```

In this case I need to define the _PhpDoc_ to return one of them (`ImplementationModel` and `AbstractModel`). If not I get another hint (I think).

Btw: You linked the wrong david ;)

**Comment by rentalhost** [comment]
I do think reasonable an inspection that detects multiple types on same variable. But maybe it should be discussed in another ticket, don't?

**Comment by funivan** [comment]
@dpauli why you cant specify interface here  `@return ImplementationModel|AbstractModel`?

**Comment by dpauli** [comment]
@funivan I can (and I will) but right now the code is not that good. I have a lot of refactoring tasks to do this.

**Comment by kalessil** [comment]
Reopening

**Comment by kalessil** [comment]
Commenting to keep stalebot away
